### DISCUSSION TITLE
Type hinting a union of callables does not work as hoped

### DISCUSSION THREAD

**Discussion by Hugovdberg** [microsoft/pylance-release#5835] [Q&A]
I'm trying to implement a robust version of the builtin `max` function by performing a reduction on an iterable with a given initial value (which eventually will be curried using the `expression` package):

```python
def max(
        it: Iterable[_a], 
        default: _a, 
        keyfunc: Callable[[_a], _SupportsLT] | Callable[[_a], _SupportsGT],
) -> _a:
    return functools.reduce(
        lambda acc, value: acc if keyfunc(acc) > keyfunc(value) else value, 
        it, 
        default,
    )
```

where `_SupportsLT` and `_SupportsGT` are identical to the protocols defined in the stdlib typeshed.

Now the problem is that Pylance raises a reportOperatorIssue on the `keyfunc(acc) > keyfunc(value)` part. This is because `keyfunc(acc)` has type `SupportsLT | _SupportsGT`, and the same for `keyfunc(value)`, and `_SupportsLT` and `_SupportsGT` cannot be compared. However, given the fact that the `keyfunc` either always returns a `_SupportsLT` or always a `_SupportsGT` the reported error condition would never occur. 

Is there a way to modify the type hints such that Pylance can infer that the comparison will always succeed?

**Comment by erictraut** [comment]
If `keyfunc(acc)` returns a value that supports `__lt__` and `keyfunc(value)` returns a value that supports `__gt__`, the function would fail. The expression `keyfunc(acc) > keyfunc(value)` would generate an exception at runtime. So I think pyright is correct in telling you that there's an error here.

**Reply by Hugovdberg** [reply]
But in my mind the type hint states that keyfunc should either be a function that turns all `a`'s in to something that supports `__lt__` or turns all `a`'s into something that supports `__gt__` (or both `__lt__` and `__gt__`), but not some `a`'s into `__lt__` and other `a`'s in `__gt__` supporting types. Since `acc` and `value` always are of the same type, the result of `keyfunc` should also always be of the same type, right?

Is it correct that the current implementation interprets `Callable[[_a], _SupportsLT] | Callable[[_a], _SupportsGT]` as equal to `Callable[[_a], _SupportsLT | _SupportsGT]`? Because that seems to me what's happening here.
